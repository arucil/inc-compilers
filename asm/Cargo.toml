[package]
name = "asm"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
indexmap = "1.8.1"
enum-iterator = "0.8.1"
num-derive = "0.3.3"
num-traits = "0.2.15"
support = { path = "../support" }
