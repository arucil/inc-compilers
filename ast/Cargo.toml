[package]
name = "ast"
version = "0.1.0"
edition = "2018"

[dependencies]
pretty = "0.10.0"
support = { path = "../support" }

[dev-dependencies]
pretty_assertions = "0.7.2"
