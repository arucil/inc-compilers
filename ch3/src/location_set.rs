use asm::{Arg, Reg};
use ast::IdxVar;
use indexmap::map::Entry;
use indexmap::IndexMap;
use num_traits::{FromPrimitive, ToPrimitive};
use smallvec::SmallVec;
use std::fmt::{self, Debug, Write};
use std::ops::Deref;

#[derive(Debug, Clone)]
pub struct LocationSet(SmallVec<[u32; 2]>);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Location(usize);

#[derive(Debug, Clone)]
pub struct VarStore(IndexMap<IdxVar, Var>);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Var(usize);

impl LocationSet {
  pub fn new() -> Self {
    Self(SmallVec::new())
  }

  pub fn add_reg(&mut self, reg: Reg) {
    let i = reg_index(reg);
    if self.0.is_empty() {
      self.0.push(0);
    }
    self.0[i / 32] |= 1 << (i % 32);
  }

  pub fn remove_reg(&mut self, reg: Reg) {
    let i = reg_index(reg);
    if !self.0.is_empty() {
      self.0[i / 32] &= !(1 << (i % 32));
    }
  }

  pub fn add_var(&mut self, var: Var) {
    let i = var_index(var);
    let offset = i / 32;
    if offset >= self.0.len() {
      self.0.resize(offset + 1, 0);
    }
    self.0[offset] |= 1 << (i % 32);
  }

  pub fn remove_var(&mut self, var: Var) {
    let i = var_index(var);
    let offset = i / 32;
    if offset < self.0.len() {
      self.0[offset] &= !(1 << (i % 32));
    }
  }

  pub fn remove_caller_saved_regs(&mut self) {
    self.remove_reg(Reg::Rax);
    self.remove_reg(Reg::Rcx);
    self.remove_reg(Reg::Rdx);
    self.remove_reg(Reg::Rsi);
    self.remove_reg(Reg::Rdi);
    self.remove_reg(Reg::R8);
    self.remove_reg(Reg::R9);
    self.remove_reg(Reg::R10);
    self.remove_reg(Reg::R11);
  }

  pub fn add_argument_regs(&mut self, arity: usize) {
    const ARGUMENT_REGS: [Reg; 6] =
      [Reg::Rdi, Reg::Rsi, Reg::Rdx, Reg::Rcx, Reg::R8, Reg::R9];

    assert!(arity <= 6);
    for reg in &ARGUMENT_REGS[..arity] {
      self.add_reg(*reg);
    }
  }

  pub fn iter(&self) -> LocationIter {
    let current = if self.0.len() > 0 { self.0[0] } else { 0 };
    LocationIter {
      set: self,
      index: 0,
      current,
    }
  }

  pub fn write(&self, f: &mut impl Write, var_store: &VarStore) -> fmt::Result {
    write!(f, "{{")?;
    let mut comma = false;
    for elem in self {
      if comma {
        write!(f, ", ")?;
      }
      comma = true;
      write!(f, "{:?}", elem.to_arg(var_store))?;
    }
    write!(f, "}}")
  }
}

pub struct LocationIter<'a> {
  set: &'a LocationSet,
  index: usize,
  current: u32,
}

impl<'a> Iterator for LocationIter<'a> {
  type Item = Location;

  fn next(&mut self) -> Option<Location> {
    while self.current == 0 {
      self.index += 1;
      if self.index >= self.set.0.len() {
        return None;
      }
      self.current = self.set.0[self.index];
    }
    let i = self.current.trailing_zeros() as usize + self.index * 32;
    self.current &= self.current - 1;
    Some(Location(i))
  }
}

impl<'a> IntoIterator for &'a LocationSet {
  type Item = Location;
  type IntoIter = LocationIter<'a>;

  fn into_iter(self) -> Self::IntoIter {
    self.iter()
  }
}

fn reg_index(reg: Reg) -> usize {
  reg.to_usize().unwrap()
}

fn var_index(var: Var) -> usize {
  var.0 + 16
}

impl VarStore {
  pub fn new() -> Self {
    Self(IndexMap::new())
  }

  pub fn get(&mut self, var: IdxVar) -> Var {
    let len = self.0.len();
    match self.0.entry(var) {
      Entry::Occupied(var) => *var.get(),
      Entry::Vacant(entry) => *entry.insert(Var(len)),
    }
  }
}

impl Deref for VarStore {
  type Target = IndexMap<IdxVar, Var>;

  fn deref(&self) -> &Self::Target {
    &self.0
  }
}

impl Var {
  pub fn to_arg(&self, var_store: &VarStore) -> Arg<IdxVar> {
    Arg::Var(var_store.get_index(self.0 - 16).unwrap().0.clone())
  }
}

impl Location {
  pub fn to_arg(&self, var_store: &VarStore) -> Arg<IdxVar> {
    if self.0 < 16 {
      Arg::Reg(Reg::from_usize(self.0).unwrap())
    } else {
      Arg::Var(var_store.0.get_index(self.0 - 16).unwrap().0.clone())
    }
  }

  pub fn from_arg(arg: Arg<IdxVar>, var_store: &mut VarStore) -> Option<Self> {
    match arg {
      Arg::Deref(reg, _) | Arg::Reg(reg) => Some(reg.into()),
      Arg::Var(var) => Some(Self(var_store.get(var).0 + 16)),
      Arg::Imm(_) => None,
      Arg::ByteReg(_) => unimplemented!(),
    }
  }

  pub fn to_reg(&self) -> Option<Reg> {
    Reg::from_usize(self.0)
  }

  pub fn to_var(&self) -> Option<Var> {
    if self.0 >= 16 {
      Some(Var(self.0 - 16))
    } else {
      None
    }
  }
}

impl From<Reg> for Location {
  fn from(reg: Reg) -> Self {
    Self(reg.to_usize().unwrap())
  }
}
