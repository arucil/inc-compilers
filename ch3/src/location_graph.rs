use crate::location_set::Location;
use indexmap::IndexMap;
use petgraph::graph::{Graph, NodeIndex as PrivateNodeIndex, UnGraph};
use std::marker::PhantomData;

pub struct LocationGraph<T> {
  graph: UnGraph<Location, ()>,
  nodes: IndexMap<Location, PrivateNodeIndex>,
  _marker: PhantomData<T>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct NodeIndex<T> {
  index: PrivateNodeIndex,
  _marker: PhantomData<T>,
}

impl<T> LocationGraph<T> {
  pub fn new() -> Self {
    Self {
      graph: Graph::new_undirected(),
      nodes: IndexMap::new(),
      _marker: PhantomData,
    }
  }

  pub fn insert_node(&mut self, loc: Location) -> NodeIndex<T> {
    let index = if let Some(&node) = self.nodes.get(&loc) {
      node
    } else {
      let node = self.graph.add_node(loc);
      self.nodes.insert(loc, node);
      node
    };
    NodeIndex {
      index,
      _marker: PhantomData,
    }
  }

  pub fn node_index(&self, loc: Location) -> Option<NodeIndex<T>> {
    self.nodes.get(&loc).map(|&index| NodeIndex {
      index,
      _marker: PhantomData,
    })
  }

  pub fn node_data(&self, node: NodeIndex<T>) -> Option<Location> {
    self.graph.node_weight(node.index).cloned()
  }

  pub fn add_edge(&mut self, n1: NodeIndex<T>, n2: NodeIndex<T>) {
    if !self.graph.contains_edge(n1.index, n2.index) {
      self.graph.add_edge(n1.index, n2.index, ());
    }
  }

  pub fn node_indices(&self) -> impl Iterator<Item = NodeIndex<T>> {
    self.graph.node_indices().map(|index| NodeIndex {
      index,
      _marker: PhantomData,
    })
  }

  pub fn neighbors(
    &self,
    node: NodeIndex<T>,
  ) -> impl Iterator<Item = NodeIndex<T>> + '_ {
    self.graph.neighbors(node.index).map(|index| NodeIndex {
      index,
      _marker: PhantomData,
    })
  }

  pub fn graph(&self) -> &UnGraph<Location, ()> {
    &self.graph
  }
}
