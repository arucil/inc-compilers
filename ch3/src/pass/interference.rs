use super::liveness::Info as OldInfo;
use crate::location_set::{Location, LocationSet, VarStore};
use asm::{Block, Instr, Program, Reg};
use ast::IdxVar;
use indexmap::{IndexMap, IndexSet};
use petgraph::dot::{Config, Dot};
use petgraph::graph::{Graph, NodeIndex, UnGraph};
use std::fmt::{self, Debug, Formatter};

pub struct Info {
  pub locals: IndexSet<IdxVar>,
  pub conflicts: IndexMap<String, InterferenceGraph>,
  pub var_store: VarStore,
}

pub struct InterferenceGraph {
  pub graph: UnGraph<Location, ()>,
  pub nodes: IndexMap<Location, NodeIndex>,
}

pub fn build_interference(
  mut prog: Program<OldInfo, IdxVar>,
) -> Program<Info, IdxVar> {
  let var_store = &mut prog.info.var_store;
  let live = &prog.info.live;
  let conflicts = prog
    .blocks
    .iter()
    .map(|(name, block)| {
      (name.clone(), build_graph(block, &live[name], var_store))
    })
    .collect();

  Program {
    info: Info {
      locals: prog.info.locals,
      conflicts,
      var_store: prog.info.var_store,
    },
    blocks: prog.blocks,
  }
}

fn build_graph(
  block: &Block<IdxVar>,
  live: &[LocationSet],
  var_store: &mut VarStore,
) -> InterferenceGraph {
  let mut graph = InterferenceGraph {
    graph: Graph::new_undirected(),
    nodes: IndexMap::new(),
  };
  for (i, instr) in block.code.iter().enumerate() {
    add_instr_edges(instr, &live[i + 1], var_store, &mut graph);
  }
  graph
}

fn add_instr_edges(
  instr: &Instr<IdxVar>,
  after: &LocationSet,
  var_store: &mut VarStore,
  graph: &mut InterferenceGraph,
) {
  let mut add = |write_loc: Location| {
    let write_loc_node = graph.get_node(write_loc);
    for after_loc in after {
      if after_loc != write_loc {
        let after_loc_node = graph.get_node(after_loc);
        graph.add_edge(write_loc_node, after_loc_node);
      }
    }
  };

  match instr {
    Instr::Add(_, dest)
    | Instr::Sub(_, dest)
    | Instr::Neg(dest)
    | Instr::Pop(dest) => {
      if let Some(dest_loc) = Location::from_arg(dest.clone(), var_store) {
        add(dest_loc);
      }
    }
    Instr::Call(_, _) => {
      add(Reg::Rax.into());
      add(Reg::Rcx.into());
      add(Reg::Rdx.into());
      add(Reg::Rsi.into());
      add(Reg::Rdi.into());
      add(Reg::R8.into());
      add(Reg::R9.into());
      add(Reg::R10.into());
      add(Reg::R11.into());
    }
    Instr::Mov(src, dest) => {
      if let Some(dest_loc) = Location::from_arg(dest.clone(), var_store) {
        let src_loc = Location::from_arg(src.clone(), var_store);
        let dest_loc_node = graph.get_node(dest_loc);
        for after_loc in after {
          if after_loc != dest_loc && Some(after_loc) != src_loc {
            let after_loc_node = graph.get_node(after_loc);
            graph.add_edge(dest_loc_node, after_loc_node);
          }
        }
      }
    }
    Instr::Push(_) | Instr::Ret | Instr::Syscall | Instr::Jmp(_) => {}
    _ => {}
  }
}

impl InterferenceGraph {
  fn get_node(&mut self, loc: Location) -> NodeIndex {
    if let Some(&node) = self.nodes.get(&loc) {
      node
    } else {
      let node = self.graph.add_node(loc);
      self.nodes.insert(loc, node);
      node
    }
  }

  fn add_edge(&mut self, n1: NodeIndex, n2: NodeIndex) {
    if !self.graph.contains_edge(n1, n2) {
      self.graph.add_edge(n1, n2, ());
    }
  }
}

impl Debug for Info {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    let graph = self.conflicts["start"]
      .graph
      .map(|_, var| var.to_arg(&self.var_store), |_, _| ());
    Dot::with_config(&graph, &[Config::EdgeNoLabel]).fmt(f)
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use asm::{Arg, Block};
  use ch2::pass::instruction::Info as OldOldInfo;
  use insta::assert_snapshot;
  use maplit::hashmap;

  fn var(name: &str) -> Arg<IdxVar> {
    Arg::Var(IdxVar {
      name: name.to_owned(),
      index: 0,
    })
  }

  #[test]
  fn example_in_book() {
    use asm::Reg::*;
    use Arg::*;
    use Instr::*;
    let code = vec![
      Mov(Imm(1), var("v")),
      Mov(Imm(42), var("w")),
      Mov(var("v"), var("x")),
      Add(Imm(7), var("x")),
      Mov(var("x"), var("y")),
      Mov(var("x"), var("z")),
      Add(var("w"), var("z")),
      Mov(var("y"), var("t")),
      Neg(var("t")),
      Mov(var("z"), Reg(Rax)),
      Add(var("t"), Reg(Rax)),
      Jmp("conclusion".to_owned()),
    ];
    let label_live = hashmap! {
      "conclusion".to_owned() => {
        let mut set = LocationSet::new();
        set.add_reg(Rax);
        set.add_reg(Rsp);
        set
      }
    };
    let prog = Program {
      info: OldOldInfo {
        locals: IndexSet::new(),
      },
      blocks: vec![("start".to_owned(), Block { code })],
    };
    let prog = super::super::liveness::analyze_liveness(prog, label_live);
    let result = build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }

  #[test]
  fn call() {
    use asm::Reg::*;
    use Arg::*;
    use Instr::*;
    let code = vec![
      Pop(Reg(Rdi)),
      Pop(Reg(Rsi)),
      Push(var("x")),
      Mov(Reg(Rbx), var("w")),
      Call("foo".to_owned(), 3),
      Add(Reg(Rax), var("w")),
      Jmp("conclusion".to_owned()),
    ];
    let label_live = hashmap! {
      "conclusion".to_owned() => {
        let mut set = LocationSet::new();
        set.add_reg(Rax);
        set.add_reg(Rsp);
        set
      }
    };
    let prog = Program {
      info: OldOldInfo {
        locals: IndexSet::new(),
      },
      blocks: vec![("start".to_owned(), Block { code })],
    };
    let prog = super::super::liveness::analyze_liveness(prog, label_live);
    let result = build_interference(prog);

    assert_snapshot!(format!("{:?}", result.info));
  }
}
