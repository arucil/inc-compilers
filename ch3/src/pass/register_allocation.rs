use super::interference::{Info as OldInfo, Interference, Moves};
use crate::location_graph::{LocationGraph, NodeIndex};
use crate::location_set::{Location, Var, VarStore};
use asm::{Arg, Block, Instr, LabelOrArg, Program, Reg};
use ast::IdxVar;
use indexmap::{IndexMap, IndexSet};
use priority_queue::PriorityQueue;
use std::cmp::Ordering;
use std::collections::{HashMap, HashSet};
use std::fmt::{self, Debug, Display, Formatter};

pub struct Info {
  pub locals: IndexSet<IdxVar>,
  /// in bytes
  pub stack_space: usize,
  pub used_callee_saved_regs: IndexSet<Reg>,
}

impl Display for Info {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    writeln!(f, "locals: {:?}", self.locals)?;
    write!(f, "used_callee_saved_regs: ",)?;
    let mut comma = false;
    for reg in &self.used_callee_saved_regs {
      if comma {
        write!(f, ", ")?;
      }
      comma = true;
      write!(f, "{}", reg)?;
    }
    writeln!(f, "\nstack_space: {} bytes", self.stack_space)
  }
}

pub fn allocate_registers(
  prog: Program<OldInfo, IdxVar>,
  available_regs: &[Reg],
) -> Program<Info> {
  assert!(prog.funs.is_empty());

  let mut num_locals = 0;
  let blocks;
  let used_callee_saved_regs;

  {
    let reg_colors: HashMap<_, _> = available_regs
      .iter()
      .enumerate()
      .map(|(i, &reg)| (reg, i as _))
      .collect();

    let mut alloc = RegisterAlloc {
      num_locals: prog.info.locals.len(),
      conflicts: &prog.info.conflicts,
      moves: &prog.info.moves,
      reg_colors: &reg_colors,
      available_regs,
      used_callee_saved_regs: IndexSet::new(),
      assign_instr_registers: gen_assign_instr_registers(
        &prog.info.var_store,
        available_regs,
        &mut num_locals,
      ),
    };

    blocks = prog
      .blocks
      .into_iter()
      .map(|block| alloc.allocate_block_registers(block))
      .collect();

    used_callee_saved_regs = alloc.used_callee_saved_regs;
  }

  Program {
    info: Info {
      locals: prog.info.locals,
      stack_space: num_locals * 8,
      used_callee_saved_regs,
    },
    blocks,
    funs: vec![],
    ..prog
  }
}

pub struct RegisterAlloc<'a, F> {
  pub num_locals: usize,
  pub conflicts: &'a LocationGraph<Interference>,
  pub moves: &'a LocationGraph<Moves>,
  pub reg_colors: &'a HashMap<Reg, u32>,
  pub available_regs: &'a [Reg],
  pub used_callee_saved_regs: IndexSet<Reg>,
  pub assign_instr_registers: F,
}

impl<'a, F> RegisterAlloc<'a, F>
where
  F: FnMut(Instr<IdxVar>, &IndexMap<Var, Color>) -> Instr,
{
  pub fn allocate_block_registers(&mut self, block: Block<IdxVar>) -> Block {
    let var_colors =
      color_graph(self.conflicts, self.moves, self.num_locals, self.reg_colors);

    for (_, c) in &var_colors {
      if let Some(&reg) = self.available_regs.get(c.0 as usize) {
        if reg.is_callee_saved() {
          self.used_callee_saved_regs.insert(reg);
        }
      }
    }

    Block {
      code: block
        .code
        .into_iter()
        .map(|instr| (self.assign_instr_registers)(instr, &var_colors))
        .collect(),
      ..block
    }
  }
}

pub fn gen_assign_instr_registers<'a>(
  var_store: &'a VarStore,
  available_regs: &'a [Reg],
  num_vars: &'a mut usize,
) -> impl (FnMut(Instr<IdxVar>, &IndexMap<Var, Color>) -> Instr) + 'a {
  let mut vars = HashSet::new();

  move |instr, var_colors| {
    let mut assign = |arg: Arg<IdxVar>| match arg {
      Arg::Deref(reg, i) => Arg::Deref(reg, i),
      Arg::Reg(reg) => Arg::Reg(reg),
      Arg::Imm(i) => Arg::Imm(i),
      Arg::Var(var) => {
        let i = var_colors[&var_store.get(var.clone())].index();
        if let Some(&reg) = available_regs.get(i) {
          Arg::Reg(reg)
        } else {
          if vars.insert(var) {
            *num_vars += 1;
          }
          let local = i - available_regs.len() + 1;
          Arg::Deref(Reg::Rbp, -8 * local as i32)
        }
      }
      Arg::ByteReg(reg) => Arg::ByteReg(reg),
    };

    match instr {
      Instr::Add { src, dest } => Instr::Add {
        src: assign(src),
        dest: assign(dest),
      },
      Instr::Call {
        label: LabelOrArg::Label(label),
        arity,
        gc,
      } => Instr::Call {
        label: LabelOrArg::Label(label),
        arity,
        gc,
      },
      Instr::Call {
        label: LabelOrArg::Arg(arg),
        arity,
        gc,
      } => Instr::Call {
        label: LabelOrArg::Arg(assign(arg)),
        arity,
        gc,
      },
      Instr::LocalJmp(label) => Instr::LocalJmp(label),
      Instr::Mov { src, dest } => Instr::Mov {
        src: assign(src),
        dest: assign(dest),
      },
      Instr::Neg(dest) => Instr::Neg(assign(dest)),
      Instr::Pop(dest) => Instr::Pop(assign(dest)),
      Instr::Push(src) => Instr::Push(assign(src)),
      Instr::Ret => Instr::Ret,
      Instr::Sub { src, dest } => Instr::Sub {
        src: assign(src),
        dest: assign(dest),
      },
      Instr::Syscall => Instr::Syscall,
      Instr::Cmp { src, dest } => Instr::Cmp {
        src: assign(src),
        dest: assign(dest),
      },
      Instr::JumpIf { cmp, label } => Instr::JumpIf { cmp, label },
      Instr::Movzx { src, dest } => Instr::Movzx {
        src: assign(src),
        dest: assign(dest),
      },
      Instr::SetIf { cmp, dest } => Instr::SetIf {
        cmp,
        dest: assign(dest),
      },
      Instr::Xor { src, dest } => Instr::Xor {
        src: assign(src),
        dest: assign(dest),
      },
      Instr::Lea { label, dest } => Instr::Lea {
        label,
        dest: assign(dest),
      },
      _ => unreachable!("{}", instr),
    }
  }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Color(u32);

#[derive(Debug, Clone, PartialEq, Eq)]
struct NodeState {
  location: Location,
  saturation: Saturation,
  num_moves: usize,
}

impl Color {
  pub fn index(&self) -> usize {
    self.0 as usize
  }
}

impl NodeState {
  fn new(location: Location, locals: usize, num_moves: usize) -> Self {
    Self {
      location,
      saturation: Saturation::new(locals),
      num_moves,
    }
  }
}

impl PartialOrd for NodeState {
  fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
    match self
      .saturation
      .count()
      .partial_cmp(&other.saturation.count())
    {
      Some(Ordering::Equal) => self.num_moves.partial_cmp(&other.num_moves),
      x => x,
    }
  }
}

impl Ord for NodeState {
  fn cmp(&self, other: &Self) -> Ordering {
    match self.saturation.count().cmp(&other.saturation.count()) {
      Ordering::Equal => self.num_moves.cmp(&other.num_moves),
      x => x,
    }
  }
}

#[derive(Debug, Clone, PartialEq, Eq)]
struct Saturation {
  colors: Vec<u32>,
  negative_colors: u32,
}

impl Saturation {
  fn new(size: usize) -> Self {
    Self {
      colors: vec![0; (size + 31) >> 5],
      negative_colors: 0,
    }
  }

  fn count(&self) -> u32 {
    let mut count = 0;
    for &x in &self.colors {
      count += x.count_ones();
    }
    count + self.negative_colors
  }

  fn contains(&self, color: Color) -> bool {
    self.colors[(color.0 >> 5) as usize] & (1 << (color.0 & 31)) != 0
  }

  fn insert(&mut self, color: Color) {
    self.colors[(color.0 >> 5) as usize] |= 1 << (color.0 & 31);
  }

  fn first_free_color(&self) -> Color {
    for i in 0..self.colors.len() {
      let x = self.colors[i].trailing_ones();
      if x < 32 {
        return Color((i << 5) as u32 + x);
      }
    }
    unreachable!("no free color")
  }
}

/// * `graph` - interference graph
/// * `moves` - move graph
/// * `locals` - a slice of all local variables in this block
/// * `reg_colors` - a map of colors of all available registers
fn color_graph(
  graph: &LocationGraph<Interference>,
  moves: &LocationGraph<Moves>,
  num_locals: usize,
  reg_colors: &HashMap<Reg, u32>,
) -> IndexMap<Var, Color> {
  let mut queue = PriorityQueue::<NodeIndex<Interference>, NodeState>::new();
  let mut assigned_nodes = HashMap::<NodeIndex<Moves>, Color>::new();

  for node in graph.node_indices() {
    let location = graph.node_data(node).unwrap();
    if location.to_reg().is_none() {
      let move_node = moves.node_index(location);
      let num_moves = move_node.map_or(0, |node| moves.neighbors(node).count());
      queue.push(node, NodeState::new(location, num_locals, num_moves));
    }
  }

  for node in graph.node_indices() {
    if graph.node_data(node).unwrap().to_reg().is_some() {
      for neighbor in graph.neighbors(node) {
        queue.change_priority_by(&neighbor, |v| {
          v.saturation.negative_colors += 1;
        })
      }
    }
    if let Some(reg) = graph.node_data(node).unwrap().to_reg() {
      if let Some(&color) = reg_colors.get(&reg) {
        for neighbor in graph.neighbors(node) {
          queue.change_priority_by(&neighbor, |v| {
            v.saturation.insert(Color(color));
          })
        }
      } else {
        for neighbor in graph.neighbors(node) {
          queue.change_priority_by(&neighbor, |v| {
            v.saturation.negative_colors += 1;
          })
        }
      }
    }
  }

  let mut var_colors = IndexMap::new();

  while let Some((node, state)) = queue.pop() {
    let mut color = state.saturation.first_free_color();

    // move biasing: prefer color that is move-related, but should not override
    // the preference for registers over stack locations.
    let color_is_reg = (color.0 as usize) < reg_colors.len();
    if let Some(node) = moves.node_index(state.location) {
      for neighbor in moves.neighbors(node) {
        if let Some(&neighbor_color) = assigned_nodes.get(&neighbor) {
          let neighbor_color_is_reg =
            (neighbor_color.0 as usize) < reg_colors.len();
          if color_is_reg == neighbor_color_is_reg
            && !state.saturation.contains(neighbor_color)
          {
            color = neighbor_color;
            break;
          }
        }
      }
      assigned_nodes.insert(node, color);
    }

    for neighbor in graph.neighbors(node) {
      queue.change_priority_by(&neighbor, |v| {
        v.saturation.insert(color);
      })
    }
    let var = state.location.to_var().unwrap();
    var_colors.insert(var, color);
  }

  var_colors
}

#[cfg(test)]
mod tests {
  use super::super::*;
  use super::*;
  use crate::location_set::LocationSet;
  use asm::Label;
  use ch2::pass::instruction_selection::Info as OldOldInfo;
  use indexmap::indexset;
  use insta::{assert_debug_snapshot, assert_snapshot};
  use maplit::hashmap;

  #[test]
  fn example_in_book() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      mov v, 1
      mov w, 42
      mov x, v
      add x, 7
      mov y, x
      mov z, x
      add z, w
      mov t, y
      neg t
      mov rax, z
      add rax, t
      jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Epilogue => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("v"),
          IdxVar::new("w"),
          IdxVar::new("x"),
          IdxVar::new("y"),
          IdxVar::new("z"),
          IdxVar::new("t")
        },
      },
      blocks,
      ..Program::default()
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let prog = interference::build_interference(prog);
    let result = allocate_registers(prog, &[Rcx]);

    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn call() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      pop rdi
      pop rsi
      push x
      mov w, rbx
      call foo, 3
      add w, rax
      jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Epilogue => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {IdxVar::new("x"),IdxVar::new("w")},
      },
      blocks,
      ..Program::default()
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let prog = interference::build_interference(prog);
    let result = allocate_registers(prog, &[Rdx, Rdi, Rsi, R8, R12]);

    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn call_no_enough_registers() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      pop rdi
      pop rsi
      push x
      mov w, rbx
      call foo, 3
      add w, rax
      jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Epilogue => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {IdxVar::new("x"),IdxVar::new("w")},
      },
      blocks,
      ..Program::default()
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let prog = interference::build_interference(prog);
    let result = allocate_registers(prog, &[Rdx, Rdi, Rsi, R8]);

    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn mov_same_variables() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      mov t, x
      add t, y
      mov z, t
      add z, w
      neg x
      neg y
      neg z
      neg w
    "#,
    );
    let label_live = hashmap! {};
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("x"),
          IdxVar::new("y"),
          IdxVar::new("z"),
          IdxVar::new("t"),
          IdxVar::new("w"),
        },
      },
      blocks,
      ..Program::default()
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let prog = interference::build_interference(prog);
    let result = allocate_registers(prog, &[Rdx, Rdi, Rsi, R8]);

    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn move_biasing_example_in_book() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
    start:
      mov v, 1
      mov w, 42
      mov x, v
      add x, 7
      mov y, x
      mov z, x
      add z, w
      mov t, y
      neg t
      mov rax, z
      add rax, t
      jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Epilogue => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldOldInfo {
        locals: indexset! {
          IdxVar::new("v"),
          IdxVar::new("w"),
          IdxVar::new("x"),
          IdxVar::new("y"),
          IdxVar::new("z"),
          IdxVar::new("t")
        },
      },
      blocks,
      ..Program::default()
    };
    let prog = liveness_analysis::analyze_liveness(prog, label_live);
    let prog = interference::build_interference(prog);
    let prog = move_biasing::build_move_graph(prog);
    let result = allocate_registers(prog, &[Rcx, Rdx, Rsi]);

    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn move_biasing_prefer_move_related_color() {
    let mut var_store = VarStore::new();
    let x = var_store.insert(IdxVar::new("x"));
    let y = var_store.insert(IdxVar::new("y"));
    let z = var_store.insert(IdxVar::new("z"));
    let w = var_store.insert(IdxVar::new("w"));
    let v = var_store.insert(IdxVar::new("v"));
    let interference = {
      let mut g = LocationGraph::new();
      let x_node = g.insert_node(Location::from(x));
      let y_node = g.insert_node(Location::from(y));
      let z_node = g.insert_node(Location::from(z));
      let w_node = g.insert_node(Location::from(w));
      let _v_node = g.insert_node(Location::from(v));
      let rax_node = g.insert_node(Location::from(Reg::Rax));
      let rsp_node = g.insert_node(Location::from(Reg::Rsp));
      g.add_edge(x_node, z_node);
      g.add_edge(y_node, w_node);
      g.add_edge(z_node, rax_node);
      g.add_edge(z_node, rsp_node);
      g.add_edge(x_node, rsp_node);
      g
    };
    let moves = {
      let mut g = LocationGraph::new();
      let x_node = g.insert_node(Location::from(x));
      let y_node = g.insert_node(Location::from(y));
      let v_node = g.insert_node(Location::from(v));
      g.add_edge(x_node, y_node);
      g.add_edge(x_node, v_node);
      g
    };
    let locals = vec![x, y, z, w, v];
    let result = color_graph(&interference, &moves, locals.len(), &hashmap! {});

    assert_debug_snapshot!(result);
  }
}
