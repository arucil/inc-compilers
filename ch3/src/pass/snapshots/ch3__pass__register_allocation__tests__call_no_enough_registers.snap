---
source: ch3/src/pass/register_allocation.rs
expression: result.to_string_pretty()
---
locals: {x.0, w.0}
used_callee_saved_regs: 
stack_space: 8 bytes
constants: {}

.start:
    pop rdi
    pop rsi
    push rdx
    mov qword [rbp - 8], rbx
    call foo
    add qword [rbp - 8], rax
    jmp .epilogue

